### PR TITLE
Multipart data values encoding

### PR DESCRIPTION
**Pull request by ambrozic**
- `None` values and lists of args are handled in `http3/multipart:iter_fields`
- Everything else that comes through being non `bytes` is cast to string
- Added tests for `multipart_encode`. Complex objects parsing into `multipart/form-data` body is same as in requests and urllib3

### CODE REVIEW

File: http3/multipart.py [hunk 1]
```diff
@@ -26,6 +26,10 @@ def render_headers(self) -> bytes:
         )
 
     def render_data(self) -> bytes:
+        if not isinstance(self.value, str):
+            if isinstance(self.value, bytes):
+                return self.value
+            self.value = str(self.value)
```
Comment (sethmlarson) [inline http3/multipart.py]: I'm not sure I'm a big fan of this behavior? Why not raise a TypeError saying what we accept and forcing users to pass values that make sense to this interface?

File: http3/multipart.py [hunk 2]
```diff
@@ -83,9 +89,11 @@ def iter_fields(data: dict, files: dict) -> typing.Iterator[Field]:
         yield FileField(name=name, value=value)
 
 
-def multipart_encode(data: dict, files: dict) -> typing.Tuple[bytes, str]:
+def multipart_encode(
+    data: dict, files: dict, boundary: bytes = None
```
Comment (sethmlarson) [inline http3/multipart.py]: imo we shouldn't expose the boundary interface publicly. What use-case is there to allowing users to control it?

File: http3/multipart.py (hunk 1 above)
Comment (ambrozic) [reply http3/multipart.py]: replied in issue post #119

File: http3/multipart.py (hunk 2 above)
Comment (ambrozic) [reply http3/multipart.py]: only reason was to make it easier to test. similar to urlib3. if it is unacceptable it can be removed.

Comment (sethmlarson) [reply http3/multipart.py]: Can be tested by mocking out `os.urandom()`. I think it should be removed as it shouldn't be a necessary interface.

File: http3/multipart.py [hunk 3]
```diff
@@ -15,9 +15,13 @@ def render_data(self) -> bytes:
...
 
 class DataField(Field):
-    def __init__(self, name: str, value: str) -> None:
+    def __init__(self, name: str, value: typing.Union[str, bytes]) -> None:
         self.name = name
         self.value = value
+        if not isinstance(self.name, str):
+            raise TypeError("Invalid type for name. Expected str.")
+        if not isinstance(self.value, (str, bytes)):
+            raise TypeError("Invalid type for value. Expected str or bytes.")
```
Comment (ambrozic) [inline http3/multipart.py]: Is there a more elegant way to write this 4 lines above?

Comment (sethmlarson) [reply http3/multipart.py]: Seems fine to me, maybe if we end up doing this a lot we'd want a function that checks the type of a value and raises a TypeError with the given message if it's not valid?

Comment (ambrozic) [reply http3/multipart.py]: Such function would be eventually useful. Type hinting could be optionally stricter in sense that could raise runtime errors.

File: http3/multipart.py [hunk 4]
```diff
@@ -73,7 +79,9 @@ def render_data(self) -> bytes:
 
 def iter_fields(data: dict, files: dict) -> typing.Iterator[Field]:
     for name, value in data.items():
-        if isinstance(value, list):
+        if value is None:
```
Comment (tomchristie) [inline http3/multipart.py]: Let's not do the "omit None values" thing - I think we should just let `None` raise an error, same as any other non-str/bytes type.

**Comment by tomchristie** [comment]
Great stuff - thank you!

**Pull request merged**